File: SubarraySumEqualsK.cpp
class Solution {
public:
int subarraySum(vector<int> &nums, int k)
{
    int n = nums.size();
    unordered_map<int, int> prefixSum;
    int sum = nums[0];
    prefixSum[sum] += 1;
    int answer = 0;

    if (sum == k)
        answer++;

    for (int i = 1; i < n; i++)
    {
        sum += nums[i];
        if (prefixSum.find(sum - k) != prefixSum.end())
            answer += prefixSum[sum - k];

                prefixSum[sum] += 1;

        if (sum == k)
        {
            answer++;
            continue;
        }
    }

    return answer;
}
};

File: SubarraySumEqualsK.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:

        sum = 0
        count = 0
        seen = defaultdict(int)

        for num in nums:
            sum += num
            
            if sum-k in seen:
                count += seen[sum-k]

            seen[sum] += 1

            if sum == k:
                count += 1
        
        return count
